### ISSUE TITLE
Comments that should not be moved between leading comma and element

### ISSUE THREAD

**Issue by brandon-leapyear** [parsonsmatt/fourmolu#72]
We have comments that separate out elements in our list, like
```hs
[ foo1
  -- bars are below here
, bar1
, bar2
, bar3
  -- bazes below here
, baz1
, baz2
]
```
but Fourmolu reformats this to
```hs
[ foo1
, -- bars below here
  bar1
, bar2
, bar3
, -- bazes below here
  baz1
, baz2
]
```
which is weird because it seems like it's commenting that specific element rather than everything after that point. Ormolu doesn't have this problem because it doesn't have the leading commas
```hs
-- ormolu
[ foo1,
  -- bars below here
  bar1,
  bar2,
  -- bazes below here
  baz1,
  baz2
]
```

**Comment by brandon-leapyear** [comment]
Also applicable for sum types
```hs
data Foo
  = Foo
  | Bar
  -- other constructors
  | Other1
  | Other 2
```

**Comment by expipiplus1** [comment]
I bumped into something similar during my leading arrow work and made it so that comments were sensibly attached. https://github.com/fourmolu/fourmolu/blob/35dbface79d722952163e0c862951cbd6ca175fc/src/Ormolu/Printer/Meat/Type.hs#L144-L145

A workaround is to add a comment after `foo1` and `bar2` like so

```haskell
x =
  [ foo1 --
  -- bars are below here
  , bar1
  , bar2
  , bar3 --
  -- bazes below here
  , baz1
  , baz2
  ]
```

**Comment by expipiplus1** [comment]
This fixes it I believe 

```haskell
diff --git a/src/Ormolu/Printer/Meat/Declaration/Value.hs b/src/Ormolu/Printer/Meat/Declaration/Value.hs
index 6443466..77e1597 100644
--- a/src/Ormolu/Printer/Meat/Declaration/Value.hs
+++ b/src/Ormolu/Printer/Meat/Declaration/Value.hs
@@ -720,7 +720,13 @@ p_hsExpr' s = \case
       TransStmtCtxt _ -> notImplemented "TransStmtCtxt"
   ExplicitList _ _ xs ->
     brackets s $
-      sep commaDel (sitcc . located' p_hsExprListItem) xs
+      getPrinterOpt poCommaStyle >>= \case
+        Leading ->
+          let ps = pure () : repeat (comma >> space)
+              xs' = zip ps xs
+              e = uncurry (\p -> sitcc . located' ((p >>) . p_hsExprListItem))
+          in sep breakpoint' e xs'
+        Trailing -> sep (comma >> breakpoint) (sitcc . located' p_hsExprListItem) xs
   RecordCon {..} -> do
     located rcon_con_name atom
     breakpointPreRecordBrace
```

**Comment by dustin** [comment]
I'd filed a bug which was pointed out to be a duplicate of this bug, so I'm going to move my initial report to a comment here:

Ideally, a comma is placed at the beginning of a line instead of the end to get a few benefits:

1. Any line can be commented out (except the first) without having to modify other lines.
2. Adding a new item anywhere (except before the first) requires only modifying that one line.
3. Lines can arbitrarily be rearranged (except for the first) with just a single cut and paste (or line swap) operation.
4. The leading comma is visually obvious so it's easier to spot mistakes (the comma also aligns with the opening `[` or `{` so you get a bit of alignment.

The current model invalidates most of these benefits.

In the example below, considering the following new undesirable states:

1. if I want to delete the first or last entry, I have to edit nearby lines.
2. If I want to delete the invalid line comment, I get a syntax error and have to edit the last entry.
3. If I want to add a new `c` entry with `-15` like the `b` example, I have to edit the next value.
4. If I want to swap two of my `c` entries to validate that my test works out of sequence, I have to do different things if I swap the `1` and `2` than if I swap the `2` and `3`.

```hs
-- got:

      let examples =
            [ -- 'a' just has one entry
              ('a', 1, 1_234_567)
            , -- 'b' just has one entry, but from before the window
              ('b', -15, 1_234_567)
            , -- 'c' has the same entry repeated
              ('c', 1, 1_234_567)
            , ('c', 2, 1_234_567)
            , ('c', 3, 1_234_567)
            , -- this comment doesn't is a lie and should be deleted
              ('c', 4, 1_234_567)
            ]

-- want:
      let examples =
            [ -- 'a' just has one entry
              ('a', 1, 1_234_567)
            -- 'b' just has one entry, but from before the window
            , ('b', -15, 1_234_567)
            -- 'c' has the same entry repeated
            , ('c', 1, 1_234_567)
            , ('c', 2, 1_234_567)
            , ('c', 3, 1_234_567)
            -- this comment doesn't is a lie and should be deleted
            , ('c', 4, 1_234_567)
            ]

-- alternatively:
      let examples =
            [ -- 'a' just has one entry
              ('a', 1, 1_234_567),
              -- 'b' just has one entry, but from before the window
              ('b', -15, 1_234_567),
              -- 'c' has the same entry repeated
              ('c', 1, 1_234_567),
              ('c', 2, 1_234_567),
              ('c', 3, 1_234_567),
              -- this comment doesn't is a lie and should be deleted
              ('c', 4, 1_234_567)
            ]
```

**Comment by pbrisbin** [comment]
I'd like to :+1: this issue but also add another scenario where this is very painful for our team.

Many of us use a toggle-comment keybinding on visually-selected lines. Take the following,

```hs
example :: [Int]
example =
  [ 1
  , 2
  , 3
  , 4
  , 6
  , 7
  , 8
  , 9
  , 10
  ]
```

Now select some elements and comment them,

```hs
example :: [Int]
example =
  [ 1
  -- , 2
  -- , 3
  -- , 4
  , 6
  , 7
  , 8
  , 9
  , 10
  ]
```

In this state, you can simply select them again and uncomment them by the same keybind. But after Fourmolu runs,

```hs
example :: [Int]
example =
  [ 1
  , -- , 2
    -- , 3
    -- , 4
    6
  , 7
  , 8
  , 9
  , 10
  ]
```

Uncommenting them is now a *very annoying* manual futzing with that moved comma.

In other words, this is more than just an aesthetic thing, it's actually very detrimental to a real workflow.

---

There seems to be a proposed fix in https://github.com/fourmolu/fourmolu/issues/72#issuecomment-860215438 is there a reason not to apply it?

**Comment by brandonchinn178** [comment]
Feel free to open a PR!

**Comment by brandonchinn178** [comment]
If it works, it's a good fix 😛

I don't see anything immediately wrong with it, so make a PR, write some tests, and see where it goes
